File: C/017.c
#include <stdio.h>

int solution(int *units, int *tens, int *hunds, int *special)
{
	int sum = 0, n;
	for (int i = 0; i < 10; i++) sum += units[i];
	n = sum;
	for (int i = 1; i < 10; i++){
		if (i != 1) sum += 10 * tens[i] + n;
		else for (int j = 0; j < 10; j++) sum += special[j];
	}
	n = sum;
	for (int i = 1; i < 10; i++) sum += 100 * hunds[i] + n - 3;
	sum += 11;
	return sum;
}


int main(void)
{
	int units[10] = {0,3,3,5,4,4,3,5,5,4};
	int tens[10]  = {0,-1,6,6,5,5,5,7,6,6};
	int hunds[10] = {0,13,13,15,14,14,13,15,15,14};
	int special[10] = {3,6,6,8,8,7,7,9,8,8};
	
	printf("%d\n",solution(units,tens,hunds,special));
	
	return 0;
}

File: Haskell/017.hs
import Data.Time

main :: IO ()
main = do 
  !startTime <- getCurrentTime
  let !result = solution
  !endTime <- getCurrentTime
  let diffTime = diffUTCTime endTime startTime
  --putStrLn ("Solution: " ++ (show result))
  --putStrLn ("Execution Time: " ++ (show diffTime))
  putStrLn $ (reverse.tail.reverse) (show diffTime)

solution :: Int
solution = sum $ map size [1..1000]

units :: Int -> Int
units n = length $ ["","one","two","three","four","five","six","seven","eight","nine"] !! n 

tens :: Int -> Int
tens n = length $ ["","","twenty","thirty","forty","fifty","sixty","seventy","eigthy","ninety"] !! n

last2 :: Int -> Int
last2 n | l == 0            = 0 
        | l > 9  && l < 20  = s + (length $ ["ten","eleven","twelve","thirteen","fourteen","fifteen","sixteen","seventeen","eighteen","nineteen"] !! (l-10))
        | l <= 9 || l >= 20 = s + (tens (l `div` 10) + units (l `mod` 10))
        where l = n `mod` 100 
              s = if n < 100 then 0 else 3

hundreds :: Int -> Int
hundreds n | n `mod` 1000 > 99    = last2 ((n `mod` 1000) `div` 100) + 7
           | otherwise = 0

thousands :: Int -> Int
thousands n | n > 999   = last2 ((n `div` 1000) `mod` 10) + 8
            | otherwise = 0

size n = thousands n + hundreds n + last2 n

